### ISSUE TITLE
Modulesd CPU usage reduction on systems with large routing tables

### ISSUE THREAD

**Issue by brmb** [wazuh/wazuh#17684]
|Wazuh version|Component|Install type|Install method|Platform|
|---|---|---|---|---|
| 4.4.1 | wazuh-modulesd | Agent | Packages/Sources | Cumulus Linux 3.7.15 |

---

### Issue:
`wazuh-modulesd` uses 1 core of CPU and spikes up to 100% utilization everytime.

Once I apply either 1 min or 1 hour internal which got introduced in this [fix](https://github.com/wazuh/wazuh/pull/7444) `<ip_update_interval>3600</ip_update_interval>`, then overall usage drops, but it spikes every 2-3 minutes (_see 2nd image below_).

![1](https://github.com/wazuh/wazuh/assets/51971082/38b487f1-8bec-4071-91b2-37ac3cf3a0ac)

![2](https://github.com/wazuh/wazuh/assets/51971082/c230770b-74f5-44ac-856e-b000c46fc0c9)

Strace of `wazuh-modulesd` process during 100% CPU utilization on Cumulus server with 28000+ routes:

![Screenshot 2023-06-28 at 14 43 22](https://github.com/wazuh/wazuh/assets/51971082/66a50c30-7940-4019-bad7-c27f0fdffda3)

### Idea:
According to [documentation](https://documentation.wazuh.com/current/user-manual/reference/internal-options.html#logcollector), `ip_update_interval` should be used only once [out_format](https://documentation.wazuh.com/current/user-manual/reference/ossec-conf/localfile.html#ossec-localfile-out-format) option is used. From what I understand, I do not use it. If thats the case - how to disable this completely in order to remove this load spikes?

**Perhaps it is possible to implement an option in which would be possible to specify maximum CPU utilization for this task if nothing else is possible?**

![3](https://github.com/wazuh/wazuh/assets/51971082/c0c87c06-394c-406a-b1d6-d6d532313e87)

As of right now - this is unreliable for me to use in crucial network systems like switches 😞

---

Related issues:
https://github.com/wazuh/wazuh/issues/7192
https://github.com/wazuh/wazuh/pull/7444

**Comment by brmb** [comment]
@vikman90 could you take a deeper look into this? 🙏🏼

**Comment by vikman90** [comment]
Hi @brmb,

In fact, the spikes we're seeing in _wazuh-modulesd_ correspond to two modules:
-** _wazuh-agentd_**: we can limit the frequency of the query with option `<ip_update_interval>` (as I see you're using).
- **_wazuh-logcollector_**: if you're not using log formatting, you can set the query to on-demand (which will result on no queries): simply set `logcollector.ip_update_interval` to `0`.

Hope it helps.
Best!

**Comment by alaunay** [comment]
Not sure about the logcollector. Doc indicates:

https://documentation.wazuh.com/current/user-manual/reference/internal-options.html#logcollector

```
0: Obtain the IP on demand.
Warning: Systems with extensive routing tables can suffer from high CPU usage.
```

I advise to use the max, 3600.

**Comment by vikman90** [comment]
@alaunay You're right. I thought that "on demand" meant that it would only query the IP at the right time to generate a log with the format `$(host_ip)` (which explained the warning you mentioned). So if there was no `<output_format>`, there should be no routing table queries.

However, I see (and have checked) that "on demand" means that _wazuh-logcollector_ will ask for the IP in each batch of file checks (every `logcollector.vcheck_files` seconds).

In summary:

- With `logcollector.ip_update_interval`=0, the IP is queried every 64 seconds (default value of `logcollector.vcheck_files`).
- With `logcollector.ip_update_interval`=3600, the IP is queried every hour.

## Proposal

I don't know if it is a bug or a blur. In any case, my proposal (to avoid these cases) is that, if logcollector.ip_update_interval=0, the IP is not queried periodically, only at the start of the agent.

**Comment by brmb** [comment]
@vikman90 @TomasTurina 

But current fix (https://github.com/wazuh/wazuh/pull/18574) means that if we set it to `0`, we disable it and server/endpoint IP does not get retrieved at all, even on agent start/restart?
> If logcollector.ip_update_internal is set to 0, the agent won't collect the host's IP.

While your suggestion was:
> In any case, my proposal (to avoid these problems) is that, if logcollector.ip_update_interval=0, the IP shall not be queried periodically, only at the start of the agent.

**Comment by vikman90** [comment]
@brmb Yes, I slightly changed the proposal when I dug into the code.

The behavior is this:
|`ip_update_interval`|Behavior|
|---|---|
|> 0|Update the value for `$(host_ip)` every `ip_update_interval` seconds (at most).|
|0|The value for `$(host_ip)` is always "0.0.0.0".|

Best.

**Comment by brmb** [comment]
@vikman90 could you explain in more details what would I lose if I disable this? `logcollector.ip_update_interval=0`

Perhaps the Wazuh Dashboard would not show enrolled agent IP addresses?
Or perhaps every new log from that server would show 0.0.0.0 IP instead of server real IP?

**Comment by vikman90** [comment]
@brmb The agent provides an internal capability to query the local IP, looking into the routing tables (just as you described).

The "client" components of this internal service are:
- The agent daemon (_wazuh-agentd_), to report the agent's IP (as shown in the Dashboard).
- The log collector (_wazuh-logcollector), to optionally enrich the logs' format (`$(host_ip)`).

Since collecting the host's IP is expensive and the agent might need that too often, both components cache the last data returned by the IP provider:

|Component|Option|
|---|---|
|Agent daemon|`<client><ip_update_interval>`|
|Log collector|`logcollector.ip_update_interval`|


So, both options work independently, and none of them interfere with each other. If you set `logcollector.ip_update_interval=0`, the agent is still able to report its IP address.

Best.

**Comment by brmb** [comment]
Amazing explanation, thanks!
